### ISSUE TITLE
[DS Project][Medical dataset] CheXpert dataset

### ISSUE THREAD

**Issue by anasedova** [knodle/knodle#295]
This work is a part of a master project DS Project @ University of Vienna

- analyzing [CheXpert dataset](https://stanfordmlgroup.github.io/competitions/chexpert/) ([paper](https://arxiv.org/pdf/1901.07031.pdf))
- bringing it to Knodle ecosystem (converting to Knodle format)
- making experiments with different weak supervision strategies within Knodle framework

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

- test set was created from validation set
- some data exploration (e.g. distribution of pathologies in the different datasets)
- function that prepares the uncertainty labels for approaches described in the paper (e.g.: U-Ones, U-Ignore, ...)

Problems and roadblocks:
- don't have access to the reports which were used to create the weak labels for the CheXpert training set -> might be problematic for creating matrix Z for the knodle input since I can't check which rules matched (I might be misunderstanding the task or the input)
- validation and test set have few observations in current split (139 and 95 respectively) and some pathologies barely occur/don't occur at all (e.g. Fracture)

Plans for the next week: 
- continue with the preprocessing: find a way to create proper input for knodle from the data

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

- wrote a class for loading and preparing the data, which should work with DataLoader
- looked at CNN examples with pytorch and ran a small one on CIFAR-10
- looked through the gradient-matching code again and ran it on the spam dataset

Problems and roadblocks:
- currently no big problems

Plans for the next week:
- try to run gradient-matching on CheXpert, go through the cleanlab code on knodle more thoroughly

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

-	created a jupyter notebook (and a matching .py file) which contains the preprocessing and some data description/download instructions

Problems and roadblocks:
-	gradient-matching code only takes sparse matrices as input, but I currently have tensors due to torchvision transformations on the images; haven’t found a solution to turn the transformed images into sparse matrices yet

-	currently user needs to specify the path to the folder where the data is stored themselves to load the data; am still thinking about whether I can make this easier

Plans for the next week:

-	finalize the jupyter notebook, start trying out gradient matching

**Comment by anasedova** [comment]
> Plans for the next week:
> 
> * finalize the jupyter notebook, start trying out gradient matching
- open a PR with data preprocessing tutorial: Jupiter notebook + .py files

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

- finalized the preprocessing tutorial (.py file and notebook) and created a pull request
- found a way to save preprocessed data with joblib (without encountering runtime error)

Problems and roadblocks:

- storing the preprocessed data results in very large joblib files (over 100 GB for training set); haven't found a way to make them smaller without reducing the image size yet

Plans for next week:

- look for ways to improve the storing of the preprocessed data
- improve preprocessing tutorial according to suggestions
- continue to look through gradient-matching code, especially the recent changes

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

- made the requested changes in the tutorial
- tried out different ways to save the preprocessed images/resulting tensors more efficiently

Problems and roadblocks:

- still haven't found a great way to store the tensors/images (splitting the training set into multiple files seems to work best right now (20 files each around 2GB))

- trying to save the preprocessed images results in different tensors when loading the images again

Plans for next week:

- try to find a better way to store images/tensors
- start making preparations for gradient matching
- make some final changes to the tutorial in case I misunderstood some of the suggestions

**Comment by anasedova** [comment]
> What was accomplished in the previous week:
> 
> * made the requested changes in the tutorial
> * tried out different ways to save the preprocessed images/resulting tensors more efficiently
> 
> Problems and roadblocks:
> 
> * still haven't found a great way to store the tensors/images (splitting the training set into multiple files seems to work best right now (20 files each around 2GB))
> * trying to save the preprocessed images results in different tensors when loading the images again
> 
> Plans for next week:
> 
> * try to find a better way to store images/tensors
> * start making preparations for gradient matching
> * make some final changes to the tutorial in case I misunderstood some of the suggestions

Lemme know when the PR is ready for me to go through it once again

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:
-	now have two new versions of storing the data: (1) the preprocessed tensors as .npz files and (2) the resized (but not normalized) images as .jpg

-	looked into image embeddings: currently obtained them using pre-trained (not fine-tuned yet) ResNet18

-	looked at IterableDatasets and tried to write a version for CheXpert

Problems and roadblocks:
-	embeddings currently just done using pre-trained (not fine-tuned) ResNet18; a different CNN might be more appropriate

-	not sure if IterableDataset really helps: seems to be mainly useful when the total length of the data isn’t known in advance (e.g. streaming data)

-	also, I might have created the ___iter___() function of the IterableDataset class wrongly (I didn’t find many examples using IterableDataset online, I mainly looked at this one: https://discuss.pytorch.org/t/using-chaindataset-to-combine-iterabledataset/85236); the way I created it doesn't seem to help with the entire training set not being storable in a single variable

Plans for the next week:
-	finish the embeddings + pull request (maybe include image embeddings in tutorial?)

**Comment by LenaZellinger** [comment]
The new notebook is now pushed.
Also, this is the link we looked at in the meeting: https://github.com/jrzech/reproduce-chexnet (about the dataloader for CheXpert)

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

•	trained a logistic regression baseline (5 epochs, performs fairly well)
•	tried gradient-matching on the data (but ran into some issues)
•	cleaned up the .py file for the embeddings and started working on a corresponding notebook
•	pushed new version of the pull request (with the corrections Andy made after running the code)

Problems and roadblocks:
•	when executing the gradient-matching code with the chexpert emebeddings, this statement in the labels_recalculation file throws an error  early in the training process (still in the first epoch):

for param in model.parameters():
            assert (torch.allclose(param.grad1.mean(dim=0), param.grad))

•	I commented the line out just so that training runs through (for 5 epochs), but the training loss didn’t seem to decrease (always close to 2), and the final performance was significantly worse than the baseline 

•	Not sure what the concrete problem is yet, but it seems like ignore_index is chosen quite often and the baseline seems to be better at detecting “No Finding” (which is what I use for other class, so maybe that is part of the issue; weak label never seems to be changed for other class)

Plans for the next week:
•	check why gradient-matching seems to not work well on the data 
•	try some different settings with other_class/all_labels and see if that improves things
•	finish the notebook for the embeddings
•	look for other ways to measure the performance

**Comment by LenaZellinger** [comment]
What was accomplished in the previous week:

- found out why gradient-matching didn't perform well (was just a small parameter error)
- added a .py file and notebook for the image embeddings (currently put it in my GM fork, unsure where else I should push it?)
- pushed GM version for chexpert (also in my GM fork)
- tried gradient-matching with binary labels (on just one pathology), but this was still with the error in the code, so the results aren't valid
- adapted the code a little for all_lables (mainly the statistics)

Problems and roadblocks:
- gradient-matching still not better than baseline (but performance is very close)
- when using all_labels, the labels_matching matrix changes size depending on batch, since it only tries the labels up to the maximum label in the batch; is this intentional?

Plans for the next week:
- prepare the presentation (maybe run some more experiments)
- maybe make the notebook a little nicer if enough time (extra info on image embeddings in general and EfficientNet)
